gapis/api/transform2/transform.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+
+	"github.com/google/gapid/gapis/api"
+)
+
+// Transform is the interface that wraps the basic Transform functionality.
+type Transform interface {
+	// BeginTransform is called before transforming any command.
+	BeginTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error)
+
+	// EndTransform is called after all commands are transformed.
+	EndTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error)
+
+	// TransformCommand takes a given command list and a state.
+	// It outputs a new set of commands after running the transformation.
+	// Transform must not modify cmd(s) in any way.
+	TransformCommand(ctx context.Context, id api.CmdID, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error)
+
+	// Melih Note: This will increase our active memory usage.
+	// Previously after every transform mutated, we were releasing memory.
+	// With the current change, it may run a few transforms until a command mutation
```

Comment:
What exactly does this mean?
How many mutations will run before resources are cleaned? If *no-one* needs up to date state, will this just defer mutations forever?

Reply:
For each command, if there are no transform that needs an accurate state, then it will do all the transforms until the MutateAndWrite then will cleanup resources